### PR TITLE
MINT-830 rework voting page with new SC

### PR DESCRIPTION
**Pull request by wAISw**
Rework redux object to model, implement voting provider (started version), implement voting manager (started version), added getters for counts

### CODE REVIEW

File: src/dao/ContractsManagerDAO.js [hunk 1]
```diff
@@ -54,9 +54,9 @@ const daoMap = {
   [DAO_EXCHANGE_MANAGER]: ExchangeManagerDAO,
   [DAO_EXCHANGE]: ExchangeDAO,
   [DAO_ERC20_MANAGER]: ERC20ManagerDAO,
-  [DAO_VOTING]: VotingDAO,
-  [DAO_VOTING_DETAILS]: VotingDetailsDAO,
-  [DAO_VOTING_ACTOR]: VotingActorDAO,
+  [DAO_VOTING_MANAGER]: VotingManagerDAO,
```
Comment (dkchv) [inline src/dao/ContractsManagerDAO.js]: actually contractManager returns only ~10 addresses, but we have about ~20 types. Ask @ahiatsevich how to sync this types this our code. May be update contractManager and add method `getContracts` with `(address[], names[])` return.

File: src/dao/VotingManagerDAO.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  IS_CREATED,
+  IS_REMOVED,
+  IS_UPDATED,
+} from 'models/notices/PollNoticeModel'
+import PollModel from 'models/PollModel'
+import ipfs from 'utils/IPFS'
+import { MultiEventsHistoryABI, VotingManagerABI } from './abi'
+import AbstractMultisigContractDAO from './AbstractMultisigContractDAO'
+import PollDetailsModel from '../models/PollDetailsModel'
+import FileModel from '../models/FileSelect/FileModel'
```
Comment (dkchv) [inline src/dao/VotingManagerDAO.js]: remove '../'

File: src/dao/VotingManagerDAO.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        files: poll.files() && poll.files(),
+        options: poll.options() && poll.options().toArray(),
+      })
+    } catch (e) {
+      // eslint-disable-next-line
+      console.error(e.message)
+    }
+
+    const timeDAO = await contractsManagerDAO.getTIMEDAO()
+    const voteLimitInTIME = poll.voteLimitInTIME()
```
Comment (dkchv) [inline src/dao/VotingManagerDAO.js]: get from store

File: src/models/voting/VotingMainModel.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  list: new Immutable.Map(),
+  pollsCount: new BigNumber(0),
+  activePollsCount: new BigNumber(0),
+  lastPoll: new BigNumber(0),
+}) {
+  pollsCount (value) {
+    return this._getSet('pollsCount', value)
+  }
+
+  list (value) {
```
Comment (dkchv) [inline src/models/voting/VotingMainModel.js]: use abstractFetchingCollection instead

File: src/models/voting/VotingMainModel.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import Immutable from 'immutable'
+import { abstractFetchingModel } from 'models/AbstractFetchingModel'
+import BigNumber from 'bignumber.js'
+
+export default class VotingMainModel extends abstractFetchingModel({
+  voteLimitInTIME: null,
+  list: new Immutable.Map(),
+  pollsCount: new BigNumber(0),
```
Comment (dkchv) [inline src/models/voting/VotingMainModel.js]: use size() from collection class

File: src/redux/voting/actions.js [hunk 6]
```diff
@@ -23,7 +37,11 @@ export const watchPoll = (notice: PollNoticeModel) => async (dispatch) => {
       dispatch(handlePollRemoved(notice.pollId()))
       break
     case IS_ACTIVATED:
+      dispatch(handlePollUpdated(notice.poll()), state.activePollsCount().plus(1))
+      break
     case IS_ENDED:
+      dispatch(handlePollUpdated(notice.poll()), state.activePollsCount().minus(1))
+      break
```
Comment (dkchv) [inline src/redux/voting/actions.js]: use `return`

File: src/redux/voting/getters.js [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
 
-  return state.isFetched
+  return state.isFetched()
     ? {
-      all: polls.length,
+      all: state.pollsCount().toString(),
       completed: polls.filter((p) => !p.poll().status()).length,
-      ongoing: polls.filter((p) => p.poll().active()).length,
-      inactive: polls.filter((p) => !p.poll().active()).length,
+      ongoing: state.activePollsCount().toString(),
```
Comment (dkchv) [inline src/redux/voting/getters.js]: do not use `BigNumber` for counters

File: src/redux/voting/reducer.js [hunk 8]
```diff
@@ -8,63 +9,45 @@ export const POLLS_REMOVE = 'voting/POLLS_REMOVE'
...
-      return {
-        ...state,
-        isFetching: false,
-        isFetched: true,
-        list: new Immutable.Map(action.list),
-      }
+      return state
+        .isFetching(false)
+        .isFetched(true)
+        .list(new Immutable.Map(action.list))
```
Comment (dkchv) [inline src/redux/voting/reducer.js]: rework with collection class

File: src/services/VotingService.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+} from '../dao/VotingManagerDAO'
+
+class VotingService extends EventEmitter {
+
+  constructor () {
+    super(...arguments)
+    this._cache = {}
+  }
+
+  getPollBackendDAO (address) {
```
Comment (dkchv) [inline src/services/VotingService.js]: all these methods will be assembled to Base class, so rename its to more generic `getDAO`, `createDAO`, `subscribeToDAO`, `unsibcribeFromDAO`. Avoid `poll`, `manager`, etc

File: src/services/VotingService.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    if (!this._cache[address]) {
+      this._cache[address] = new PollBackendDAO(address)
+    }
+    return this._cache[address]
+  }
+
+  getVotingManager () {
+    if (!this._cache['VotingManager']) {
+      this._cache['VotingManager'] = new VotingManagerDAO()
```
Comment (dkchv) [inline src/services/VotingService.js]: manager is not a part of these cache, use `this._manager`

Comment (dkchv) [reply src/services/VotingService.js]: init manager in constructor

File: src/services/VotingService.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    return this._cache['VotingManager']
+  }
+
+  subscribeToPoll (address) {
+    if (this._cache[address]) return null
```
Comment (dkchv) [inline src/services/VotingService.js]: split to new line

File: src/models/voting/VotingMainModel.js (hunk 4 above)
Comment (wAISw) [reply src/models/voting/VotingMainModel.js]: it's list of polls. list = new VotingCollection(). It's ok.

File: src/redux/voting/actions.js (hunk 6 above)
Comment (wAISw) [reply src/redux/voting/actions.js]: no, because after switch we have one more dipatch

File: src/components/dialogs/VoteDialog.jsx [hunk 12]
```diff
@@ -136,21 +149,23 @@ export default class VoteDialog extends PureComponent {
                     <div styleName='entryValue'>
                       {details.voteLimitInTIME === null
                         ? <i>Unlimited</i>
-                        : <span>{details.voteLimitInTIME.toString()} TIME</span>
+                        : <span>{timeToken.removeDecimals(details.voteLimitInTIME).toString()} TIME</span>
```
Comment (dkchv) [inline src/components/dialogs/VoteDialog.jsx]: use Amount and TokenValue in these and appropriate components

Comment (dkchv) [reply src/components/dialogs/VoteDialog.jsx]: convert values on fetch

File: src/components/dialogs/VoteDialog.jsx [hunk 13]
```diff
@@ -136,21 +149,23 @@ export default class VoteDialog extends PureComponent {
...
                   <div styleName='entry entryReceived'>
                     <div styleName='entryLabel'><Translate value={prefix('receivedVotes')} />:</div>
-                    <div styleName='entryValue'>{details.received.toString()} TIME</div>
+                    <div styleName='entryValue'>{timeToken.removeDecimals(details.received).toString()} TIME</div>
                   </div>
                   <div styleName='entry entryVariants'>
                     <div styleName='entryLabel'><Translate value={prefix('variants')} />:</div>
-                    <div styleName='entryValue'>{details.options.count() || (<i><Translate value={prefix('no')} /></i>)}</div>
+                    <div styleName='entryValue'>{details.options.count() || (
+                      <i><Translate value={prefix('no')} /></i>)}</div>
```
Comment (dkchv) [inline src/components/dialogs/VoteDialog.jsx]: add one `const` for this node

File: src/dao/PollEmitterDAO.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
+import contractsManagerDAO from 'dao/ContractsManagerDAO'
+import PollNoticeModel, { IS_VOTED, IS_ACTIVATED, IS_ENDED } from 'models/notices/PollNoticeModel'
+import { MultiEventsHistoryABI, PollEmitterABI } from './abi'
+import AbstractMultisigContractDAO from './AbstractMultisigContractDAO'
+
+export const EVENT_POLL_VOTED = 'PollVoted'
+export const EVENT_POLL_ACTIVATED = 'PollActivated'
+export const EVENT_POLL_ENDED = 'PollEnded'
+
+export default class PollEmitterDAO extends AbstractMultisigContractDAO {
```
Comment (dkchv) [inline src/dao/PollEmitterDAO.js]: multisig?

File: src/dao/VotingManagerDAO.js [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+export const EVENT_POLL_UPDATED = 'PollUpdated'
+export const EVENT_POLL_REMOVED = 'PollRemoved'
+
+export default class VotingManagerDAO extends AbstractMultisigContractDAO {
+  constructor (at) {
+    super(VotingManagerABI, at, MultiEventsHistoryABI)
+  }
+
+  async getVoteLimit () {
+    return await this._call('getVoteLimit')
```
Comment (dkchv) [inline src/dao/VotingManagerDAO.js]: remove await/async, add `func(): Promise`

**Pull request merged**